cpp/ql/src/experimental/Security/CWE/CWE-788/AccessOfMemoryLocationAfterEndOfBufferUsingStrlen.ql

```diff
@@ -0,0 +1,23 @@
+/**
+ * @name Access Of Memory Location After End Of Buffer
+ * @description The expression `buffer [strlen (buffer)] = 0` is potentially dangerous, if the variable `buffer` does not have a terminal zero, then access beyond the bounds of the allocated memory is possible, which will lead to undefined behavior.
+ *              If terminal zero is present, then the specified expression is meaningless.
+ * @kind problem
+ * @id cpp/access-memory-location-after-end-buffer
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-788
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+from StrlenCall fc, AssignExpr expr, ArrayExpr exprarr
+where
+  exprarr = expr.getLValue() and
+  expr.getRValue().getValue().toInt() = 0 and
+  exprarr.getArrayOffset() = fc and
```

Comment:
This is matching _specifically_ on `arr[strlen(arr)] = 0`, so it doesn't match this situation:
```codeql
int length = strlen(arr);
arr[length] = 0;
```

If you instead replaced this line with:
```codeql
globalValueNumber(exprarr.getArrayOffset()) = globalValueNumber(fc)
```
you would catch this pattern as well.

Reply:
Thank you for your offer.
it expands the detection.
unfortunately, false positives also appear.
I decided to limit them, but it turned out not very elegant.
therefore I would be glad to hear your opinion.

Reply:
I'm surprised this causes false positives. Can you give an example of this? Or better: Can you add a false positive that is prevented to your [test.c](https://github.com/github/codeql/blob/b7df18b97e75879ffed2f2c430072c05eba43584/cpp/ql/test/experimental/query-tests/Security/CWE/CWE-788/semmle/tests/test.c) file?

Reply:
examples of false detection.  
https://lgtm.com/projects/g/pure-data/pure-data/snapshot/70079c797a95eb85df1d00a3409b262eefbf7227/files/src/g_numbox.c?sort=name&dir=ASC&mode=heatmap#x11da8baeead003d6:1  
https://lgtm.com/projects/g/orestis/python/snapshot/6c68ef04be7f8ae5f07b3d1f11e8b9578633ca6f/files/Python/import.c?sort=name&dir=ASC&mode=heatmap#x92b40dcfed6bdfe9:1  
https://lgtm.com/projects/g/orestis/python/snapshot/6c68ef04be7f8ae5f07b3d1f11e8b9578633ca6f/files/Modules/getpath.c?sort=name&dir=ASC&mode=heatmap#x12027adcd4ed831e:1  
https://lgtm.com/projects/g/orestis/python/snapshot/6c68ef04be7f8ae5f07b3d1f11e8b9578633ca6f/files/Modules/getpath.c?sort=name&dir=ASC&mode=heatmap#x7bc1e30a68df9f29:1  

I can't add them to `test.c` yet, since yesterday's yesterday my codeql cli layout has stopped correctly running tests. I have not yet figured out the reason for the breakdown.

Reply:
Oh. Yeah, those false positives make sense. Good catch!